content/ADR-162-auth-chain-with-permissions.md

```diff
@@ -0,0 +1,311 @@
+on behalf of the user, such as scenes in the Catalyst Network or a scene for
+a World using a DCL name.
+
+Following the principle of the least privilege it would be useful to be able
+to grant the ephemeral key only the set of permissions that are required
+for the purpose of the client using it.
+
+* In case of Explorer, just deploying profiles should be enough.
+* When using `dcl deploy` the only permissions
+  required are for deploying scenes (or wearables? emotes?). Not needed for
+  deploying profiles. However, in this case, no ephemeral key is used but
+  rather the deployment is signed directly with the user wallet.
+
+So this proposal is about extending the Auth Chain to support the creation
+of the ephemeral key with a given set of permissions associated to it, and
+that the server that receives those requests can validate that the user granted
+access to the scopes required for the operation being requested. This way, the
+user could sign the ephemeral key to only allow deployment of worlds under a
+certain name but not deployment of profiles nor scenes in the Catalyst network.
+
+It is worth mentioning that this does not have any effect or implications on
+Auth Chains in which there is no ephemeral key, but rather the user signs the
+request directly. This is usually the case for scene deployments. For e.g.:
+
+```json
+{
+  "version": "v3",
+  "localTimestamp": 1669232361483,
+  "authChain": [
+    {
+      "type": "SIGNER",
+      "payload": "0xe2b6024873d218b2e83b462d3658d8d7c3f55a18",
+      "signature": ""
+    },
+    {
+      "type": "ECDSA_SIGNED_ENTITY",
+      "payload": "bafkreignljg5bvmzczke42gymktbraf7py7riwyclmbgzmwcyswxdgktju",
+      "signature": "0x82ccde2c7c6b300566c40fd6f3234876614564a6e13643e968fe4f69828a2fb41e8286fbf94ac92a19a5dfb96ff636a2d5e41406d5dfc200e76145cc4b0d96321c"
+    }
+  ]
+}
+```
+
+## Solution Space Exploration
+
+### Permission specification
+
+Permissions must be specified when creating an ephemeral key. A simple and very
+flexible way for this could be similar to how Amazon does it
+in [AWS IAM](https://aws.amazon.com/iam/). In this case, it would consist of
+a collection of a resource name and either an allow or deny expression.
+
+For e.g.:
+
+```yaml
+Permissions:
+  - resource: "my-dcl-name.dcl.eth"
+    allow: "dcl:worlds:deploy"
+  - resource: "0,0"
+    allow: "dcl:scene:deploy"
+  - resource: "0xaddress"
+    allow: "dcl:explorer:*"
+  - resource: "0xaddress"
+    deny: "dcl:explorer:denied-operation"
+```
+
+The `resource` name must be an item referencing an entity of the domain that
+the server knows. For e.g. for DCL Worlds, it can be a DCL name, while for
+scene deployment in Genesis City it can be the coordinates of the land, or
+for a profile deployment it can be the profile address.
+
+The `allow` and `deny` elements are structured as follows:
+`namespace:service-name:operation-name` or `namespace:service-name:*`.
```

Comment:
Could those elements be anything or would it be a list of known operations? 

Thinking in bad actors and the case of _denying_ misspelled operations and users thinking it is OK. For example,
let's say we have the operation `dcl:explorer:deploy` but some signs this:
```  
- resource: "0xaddress"
  allow: "dcl:explorer:*"
- resource: "0xaddress"
  deny: "dcl:explorer:deployment"
```
but it actually gets `dcl:explorer:deploy` operation allowed. Could this be a problem?

Also, by `least privilege` I understand that all operations would be _denied_ by default, is that right?

Reply:
It's kind of stale right now, there are a few things that already need re-thinking.
Happy to open up again for discussion / work on it.